src/ai/backend/agent/docker/kernel.py

```diff
@@ -149,31 +151,26 @@ async def commit(self, path: Path):
         filename = f"{now}.tar.gz"
         filepath = path / filename
         try:
-            Path(filepath).parent.mkdir(exist_ok=True, parents=True)
+            Path(path).mkdir(exist_ok=True, parents=True)
         except ValueError:  # parent_path does not start with work_dir!
             raise AssertionError("malformed committed path.")
-        loop = current_loop()
         lock_path = self._get_lock_path(path)
         try:
             async with FileLock(path=lock_path, timeout=0.1):
-                docker = Docker()
                 log.info("Container is being committed to {}", filepath)
-                async with docker._query(
-                    f"containers/{container_id}/export",
-                    method="GET",
-                ) as response:
-                    log.debug("Container commit data is being written on {}", filepath)
-                    data: bytes = await response.read()
-
-                    def _save_tar():
-                        tar_info = tarfile.TarInfo(filename)
-                        tar_info.size = len(data)
-                        with tarfile.open(name=filepath, mode="x:gz") as tar:
-                            tar.addfile(tar_info, fileobj=BytesIO(data))
-
-                    await loop.run_in_executor(None, _save_tar)
+                docker = Docker()
+                container = docker.containers.container(container_id)
+                response: Mapping[str, Any] = await container.commit()
+                image_id = response["Id"]
+                async with docker._query(f"images/{image_id}/get") as tb_resp:
+                    with tarfile.open(str(filepath), "w|gz") as tar:
+                        async for chunk in tb_resp.content.iter_chunked(DEFAULT_CHUNK_SIZE):
+                            assert tar.fileobj is not None
+                            tar.fileobj.write(chunk)
```

Comment:
This should reside in a separate thread. Simply we could use `asyncio.run_in_executor()`, or employ the structure of `common.files.AsyncFileWriter` which uses `janus` to stream the chunks via a queue shared by an executor task and the asyncio world.
For now, let's wrap it up using `asyncio.run_in_executor()` for simpler implementation.

Reply:
but the agent can process multiple requests at the same time with this code.
it can check the duplicated(or on-going) session commit and can create new session while the file write streaming is ongoing.

besides, I am not sure how to implement `asyncio.run_in_executor()`. since `tarfile` does not support appending mode for opening file.

Reply:
I will fix it by myself this part.